src/main/java/org/zeromq/ZMQ.java

```diff
@@ -3173,6 +3174,27 @@ public int sendByteBuffer(ByteBuffer data, int flags)
             return Stream.generate(this::recv);
         }
 
+        /**
+         * This API is in DRAFT state and is subject to change at ANY time until declared stable
+         * handle incoming message with a handler
+         *
+         * @param handler handler to handle incoming message
+         * @param flags   either:
+         *                <ul>
+         *                <li>{@link org.zeromq.ZMQ#DONTWAIT DONTWAIT}:
+         *                Specifies that the operation should be performed in non-blocking mode.
+         *                If there are no messages available on the specified socket,
+         *                the method shall fail with errno set to EAGAIN and return null.</li>
+         *                <li>0 : receive operation blocks until one message is successfully retrieved,
+         *                or stops when timeout set by {@link #setReceiveTimeOut(int)} expires.</li>
+         *                </ul>
+         */
+        @Draft
+        public void recv(Consumer<ZMsg> handler, int flags)
+        {
+            handler.accept(ZMsg.recvMsg(this, flags));
```

Comment:
I would rather have preferred the opposite: a static method in ZMsg with a consumer parameter. ZMQ.Socket is ZMsg-agnostic, and this starts to introduce a dependency. As far as I can see, the logic of ZMQ has been to stack API on top of each other, so users are free to choose at which level they want to operate.

Reply:
I tried to correct the placement of the API, I wish if there is a documentation about this layering, it's would be great if we have an example, implemented in various layers of the API.